### PR TITLE
Added foreach_frac API

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47385 Added foreach_trunc, foreahc_reciprocal, foreach_sigmoid APIs
* **#47384 Added foreach_frac API**
* #47875 Adding bunch of unary foreach APIs

Differential Revision: [D24737052](https://our.internmc.facebook.com/intern/diff/D24737052)

Implementing foreach_frac API.

### Planned list of ops 
- [x]  abs
- [x]  acos
- [x]  asin
- [x]  atan
- [x]  ceil
- [x]  cos
- [x]  cosh
- [x]  erf
- [x]  erfc
- [x]  exp
- [x]  expm1
- [x]  floor
- [x]  log
- [x]  log10
- [x]  log1p
- [x]  log2
- [x]  frac
- [x]  neg
- [ ]  reciprocal
- [x]  round
- [ ]  rsqrt
- [ ]  sigmoid
- [x]  sin
- [x]  sinh
- [x]  sqrt
- [x]  tan
- [x]  tanh
- [ ]  trunc
- [x]  lgamma
- [ ]  digamma
- [ ]  erfinv
- [ ]  sign
- [ ]  mvlgamma
- [ ]  clamp
- [ ]  clamp_min
- [ ]  clamp_max

### Perf results
```
----------------- OP:  frac  -----------------
  Median: 1.15 ms
  296.01 us

```

### Script

```

import torch
import torch.optim as optim
import torch.nn as nn
import torchvision
import torch.utils.benchmark as benchmark_utils

inputs = [torch.rand(3, 200, 200, device="cuda") for _ in range(100)]

def main():
    for op in [
            "frac"
        ]:
        print("\n\n----------------- OP: ", op, " -----------------")
        stmt = "[torch.{op}(t) for t in inputs]"
        timer = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer)",
        )
        print(f"autorange:\n{timer.blocked_autorange()}\n\n")


        stmt = "torch._foreach_{op}(inputs)"
        timer_mta = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer_mta)",
        )
        print(f"autorange:\n{timer_mta.blocked_autorange()}\n\n")

if __name__ == "__main__":
    main()

```

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 1]
```diff
@@ -287,6 +287,48 @@ struct UnaryOpFunctor {
         }
 };
 
+template<typename T, int depth, int r_args_depth, int res_arg_index>
+struct FracFunctor {
```
Comment (zou3519) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:291]: Why is this different from UnaryOpFunctor?

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 2]
```diff
@@ -349,4 +352,58 @@ void foreach_tensor_round_cuda_(TensorList tensors) {
     foreach_unary_op_<Round>(tensors);
 }
 
+template<typename T>
+struct Trunc {
+    __device__ T operator()(T t) const { return std::trunc(t); }
+};
+
+std::vector<Tensor> foreach_tensor_frac_cuda(TensorList tensors) {
+    check_foreach_api_restrictions(tensors);
```
Comment (zou3519) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:361]: The c++ style guide that most of the codebase follows uses 2-space indents. But as long as everything's consistent 4-space indents seem fine

**Comment by zou3519** [comment]
Btw, the PR body has timings for a lot of operators other than frac. I was confused for a bit on what exactly this PR is supposed to do

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu (hunk 2 above)
Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachUnaryOp.cu:361]: Yeah, everything is consistent

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 1 above)
Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:291]: Please, see line 313. Im following the implementation from [here](https://github.com/pytorch/pytorch/blob/9ea7a6c7c55484b83b8cedb58158f429425fe950/aten/src/ATen/native/cuda/UnaryFractionKernels.cu#L35)

Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:291]: Its worth pointing out that im planning to refactor this file once again and pass line 313 into some generic block of code.

Comment (zou3519) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:291]: So the op being passed is the Trunc struct:
```
template<typename T>
struct Trunc {
    __device__ T operator()(T t) const { return std::trunc(t); }
};
```
Why not create a struct that computes frac (via t - std::trunc(t)) and pass that into UnaryOpFunctor?

**Review by zou3519 — APPROVED** [review]
Approving to unblock, but we should really be able to refactor this

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 1 above)
Comment (izdeby) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh]: @zou3519, you are right. rewrote to use UnaryOpFunctor. FracFunctor is gone
